src/transformers/generation/utils.py

```diff
@@ -1297,6 +1297,10 @@ def _prepare_generation_config(
 
     def _get_initial_cache_position(self, input_ids, model_kwargs):
         """Calculates `cache_position` for the pre-fill stage based on `input_ids` and optionally past length"""
+        if "use_cache" in model_kwargs and not model_kwargs["use_cache"]:
```

Comment:
```suggestion
        if not model_kwargs.get("use_cache", True):
```

Reply:
If you are not using the cache, you should still pass the cache positions, but they should be automatically created using the shape of the input ids. cc @gante as well!

Reply:
Included the `use_cache` check into `_update_model_kwargs_for_generation` function.

Reply:
It seems to me that [these](https://github.com/huggingface/transformers/blob/1e05671d213642a67dd418c13a18a17a18c25117/src/transformers/models/llama/modeling_llama.py#L1084-L1086) lines should be an "if...else", which will solve the problem with caching. 

Because if sequence length is not 1, making a diagonal mask will be enough, and the `causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)` is unnecessarily unmasking all the positions now.

Reply:
@zucchini-nlp , yes, using `cache_position` when `use_cache=False` does not make much sense. But with the proposed fixes, `cache_position` is None, and it is initialized in the model's forward pass as `cache_position = torch.arange(past_seen_tokens, past_seen_tokens + inputs_embeds.shape[1], device=inputs_embeds.device)`, therefore it does not influence the update of `causal_mask`.